model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java

```diff
@@ -174,7 +174,9 @@ public void addFederatedIdentity(RealmModel realm, UserModel user, FederatedIden
     @Override
     public void updateFederatedIdentity(RealmModel realm, UserModel federatedUser, FederatedIdentityModel federatedIdentityModel) {
         FederatedIdentityEntity federatedIdentity = findFederatedIdentity(federatedUser, federatedIdentityModel.getIdentityProvider(), LockModeType.PESSIMISTIC_WRITE);
+        if (federatedIdentity == null) return;
 
+        federatedIdentity.setUserId(federatedIdentityModel.getUserId());
```

Comment:
Do we need to update the user ID? Why the ID will change during authentication?

I'm wondering if that will make it possible to take over accounts if you manage to somehow update the user ID after authenticating as some other user.

Reply:
Yes.
These are possible two different broker users when updating.
I think we should update the ID as well.

If we don't update the ID, the next time login via the same IDP user will continue the first-broker-login flow.
Because here can not find the federated user based on the IDP user ID.
https://github.com/keycloak/keycloak/blob/70e15bdaa49a8babc5dd76882ca5c66462e548fe/services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java#L536-L536



> I'm wondering if that will make it possible to take over accounts if you manage to somehow update the user ID after authenticating as some other user.

Before update all fields of the broker link, **it makes sure the `username` or `userId` is matched**.
There are two cases when updating:
case 1:
```
[before]
User A:
 - username: user_a
 - broker_user: 
     - username: user_a
     - user_id: id_1

when the new broker_link comes
(This happens when the broker_user create a new account but use the previous username)
broker_user:
    - username: user_a
    - user_id: id_2

[after]
User A:
 - username: user_a
 - broker_user: 
     - username: user_a
     - user_id: id_2
```

case 2:
```
[before]
User A:
 - username: user_a
 - broker_user: 
     - username: whatever
     - user_id: id_1

when the new broker_link comes
(This happens when the broker_user just changed its username)
broker_user:
    - username: user_b
    - user_id: id_1

[after]
User A
 - username: user_a
 - broker_user: 
     - username: user_b
     - user_id: id_1
```

The case 2 makes sure the ID not changed.
And I am not sure should we handle the case 1 to update the entire broker link.

Please help to check.

Reply:
I see. True, I did not consider this.

My concern here is basically to make sure for the user that he is authenticating from a different account at the IdP. Otherwise, we can (not 100% sure) enter a situation where an attacker can somehow trick the user into authenticating at some random IdP and have their account linked to the attacker. 

To be safe, perhaps we should introduce a step during the flow to say "Hey, looks like you have an existing account linked to Foo IdP but using a different than the account you are now authenticating. Do you confirm using the new account instead?"

Something like that. I hope it makes sense.

Reply:
@pedroigor From what I recall, the problem arises post authentication via 'IdpConfirmLinkAuthenticator' (which pertains to the screen intended for user display). In my opinion, this issue seems to be accounted for already.

Regarding the scenario you've highlighted, it's plausible only if an attacker gains access to and can manipulate identities within the external IDP (already) linked to the Keycloak user account. It means an attacker cannot trigger this behavior from another random IDP, it has to be the same that is already linked to the keycloak account. In such a high-level compromise, the attacker could effectively impersonate anyone from the external IDP anyhow.

Reply:
> To be safe, perhaps we should introduce a step during the flow to say "Hey, looks like you have an existing account linked to Foo IdP but using a different than the account you are now authenticating. Do you confirm using the new account instead?"

@pedroigor 
Agree, this is what we added in our production, which verifying the code via the email sent to the Keycloak user.

> the problem arises post authentication via 'IdpConfirmLinkAuthenticator' (which pertains to the screen intended for user display).

I think the confirming from `IdpConfirmLinkAuthenticator` is not safe enough.

Should we add such verification for the case 1 I mentioned above?
At that case, the attacker use the same username login from the IdP, it could override the broker link automatically without the real user noticing.
Or maybe it is difficult get the same username from the IdP for that he needs find the username of the user.

Reply:
Yer ... Making such a critical change automatically is not a good thing. An additional step should at least help users to identify some weird behavior if they are (and perhaps most of the time) authenticating using a same account at the IdP.

Together with the recent events that were added by you, you can build a more robust notification system when this happens.

So yeah, I would vote for introducing this additional step.

Reply:
@lexcao @pedroigor I wonder if it makes sense to introduce some new authenticator to the `First Broker Login` flow, which will show this confirmation screen and set some attribute to the broker context? Then updating the broker link would be possible just when this attribute is present in the broker context (where it was previously set by this authenticator). Or something along those lines?

I am just thinking about this due the:
- It might be better to introduce authenticator rather than new endpoint to `IdentityBrokerService` and all the infrastructure around this
- Will be good as this behaviour would be flexible and possibly not enabled for all the deployments. As based on the comments like for example https://github.com/keycloak/keycloak/pull/20677#issuecomment-1625978775 and

Reply:
@mposolda Sounds like a plan. But do you want to have it now in this PR or as a follow-up?

Reply:
I guess that now it would be easier as otherwise we would need to introduce new endpoint(s) to `IdentityBrokerService` for the whole shebang around showing custom screen with text "Hey, looks like you have an existing account linked to Foo IdP but using a different than the account you are now authenticating. Do you confirm using the new account instead?" if I understand correctly? But when using authenticator SPI, there won't be much customizations needed in the `IdentityBrokerService` .

Reply:
@mposolda 
Sounds a good idea, I prefer adding a new first login flow Authenticator for this case rather than customizing `IdentityBrokerService`.

Maybe we need split into a new issue to track this and I will create a new PR to implement such feature,
close this one as well.

WDYT @pedroigor

Reply:
@lexcao +1 from my point of view

Reply:
Works for me too.

Reply:
Great, @pedroigor @mposolda 
I have created a new issue tracking this.
Further discussion will be there. 
We can close this as well.

Thank you.
https://github.com/keycloak/keycloak/issues/26201